File: packages/sdk-dotnet/src/Rollgate.SDK/Evaluate.cs
using System.Globalization;
using System.Security.Cryptography;
using System.Text;
using System.Text.RegularExpressions;

namespace Rollgate.SDK;

public static class Evaluate
{
    /// <summary>
    /// Determines if a user is in the rollout percentage using SHA-256 consistent hashing.
    /// hash = SHA-256(utf8("{flagKey}:{userId}"))
    /// value = BigEndian.Uint32(hash[0:4]) % 100
    /// result = value &lt; percentage
    /// </summary>
    public static bool IsInRollout(string flagKey, string userId, int percentage)
    {
        var input = Encoding.UTF8.GetBytes($"{flagKey}:{userId}");
        var hash = SHA256.HashData(input);
        uint value = (uint)((hash[0] << 24) | (hash[1] << 16) | (hash[2] << 8) | hash[3]);
        return (value % 100) < (uint)percentage;
    }

    public static bool MatchesCondition(string op, object? attrValue, string condValue, UserContext? user)
    {
        bool exists = attrValue != null && attrValue.ToString() != "";

        switch (op)
        {
            case "is_set": return exists;
            case "is_not_set": return !exists;
        }

        if (!exists) return false;

        var value = ToString(attrValue).ToLowerInvariant();
        var cv = condValue.ToLowerInvariant();

        return op switch
        {
            "equals" or "eq" => value == cv,
            "not_equals" or "neq" => value != cv,
            "contains" => value.Contains(cv),
            "not_contains" => !value.Contains(cv),
            "starts_with" => value.StartsWith(cv),
            "ends_with" => value.EndsWith(cv),
            "in" => SplitAndTrim(condValue).Any(v => v.ToLowerInvariant() == value),
            "not_in" => !SplitAndTrim(condValue).Any(v => v.ToLowerInvariant() == value),
            "greater_than" or "gt" => CompareNumeric(attrValue, condValue, ">"),
            "greater_equal" or "gte" => CompareNumeric(attrValue, condValue, ">="),
            "less_than" or "lt" => CompareNumeric(attrValue, condValue, "<"),
            "less_equal" or "lte" => CompareNumeric(attrValue, condValue, "<="),
            "regex" => TryRegex(ToString(attrValue), condValue),
            "semver_gt" => CompareSemver(ToString(attrValue), condValue, ">"),
            "semver_lt" => CompareSemver(ToString(attrValue), condValue, "<"),
            "semver_eq" => CompareSemver(ToString(attrValue), condValue, "="),
            _ => false
        };
    }

    public static object? GetAttributeValue(string attribute, UserContext? user)
    {
        if (user == null) return null;
        return attribute switch
        {
            "id" => user.Id,
            "email" => user.Email,
            _ => user.Attributes.TryGetValue(attribute, out var val) ? val : null
        };
    }

    public static string ToString(object? v)
    {
        if (v == null) return "";
        if (v is string s) return s;
        if (v is bool b) return b ? "true" : "false";
        if (v is double d) return d.ToString("G", CultureInfo.InvariantCulture);
        if (v is float f) return f.ToString("G", CultureInfo.InvariantCulture);
        if (v is int i) return i.ToString();
        if (v is long l) return l.ToString();
        if (v is System.Text.Json.JsonElement je)
        {
            return je.ValueKind switch
            {
                System.Text.Json.JsonValueKind.String => je.GetString() ?? "",
                System.Text.Json.JsonValueKind.Number => je.GetDouble().ToString("G", CultureInfo.InvariantCulture),
                System.Text.Json.JsonValueKind.True => "true",
                System.Text.Json.JsonValueKind.False => "false",
                _ => je.ToString()
            };
        }
        return v.ToString() ?? "";
    }

    private static string[] SplitAndTrim(string s)
        => s.Split(',').Select(p => p.Trim()).ToArray();

    private static bool CompareNumeric(object? attrVal, string condVal, string op)
    {
        if (!TryToDouble(attrVal, out var a)) return false;
        if (!double.TryParse(condVal, NumberStyles.Float, CultureInfo.InvariantCulture, out var b)) return false;
        return op switch
        {
            ">" => a > b,
            ">=" => a >= b,
            "<" => a < b,
            "<=" => a <= b,
            _ => false
        };
    }

    private static bool TryToDouble(object? v, out double result)
    {
        result = 0;
        if (v == null) return false;
        if (v is double d) { result = d; return true; }
        if (v is float f) { result = f; return true; }
        if (v is int i) { result = i; return true; }
        if (v is long l) { result = l; return true; }
        if (v is string s) return double.TryParse(s, NumberStyles.Float, CultureInfo.InvariantCulture, out result);
        if (v is System.Text.Json.JsonElement je && je.ValueKind == System.Text.Json.JsonValueKind.Number)
        {
            result = je.GetDouble();
            return true;
        }
        return false;
    }

    private static bool TryRegex(string value, string pattern)
    {
        try { return Regex.IsMatch(value, pattern); }
        catch { return false; }
    }

    private static bool CompareSemver(string attrVal, string condVal, string op)
    {
        var a = ParseVersion(attrVal);
        var b = ParseVersion(condVal);
        if (a == null || b == null) return false;

        while (a.Count < b.Count) a.Add(0);
        while (b.Count < a.Count) b.Add(0);

        for (int i = 0; i < a.Count; i++)
        {
            if (a[i] > b[i]) return op is ">" or ">=";
            if (a[i] < b[i]) return op is "<" or "<=";
        }
        return op is "=" or ">=" or "<=";
    }

    private static List<int>? ParseVersion(string v)
    {
        var clean = v.TrimStart('v');
        var parts = clean.Split('.');
        var result = new List<int>();
        foreach (var p in parts)
        {
            if (!int.TryParse(p, out var n)) return null;
            result.Add(n);
        }
        return result;
    }
}

File: packages/sdk-flutter/lib/src/evaluate.dart
import 'dart:convert';
import 'dart:typed_data';
import 'package:crypto/crypto.dart';

import 'user_context.dart';

/// Determines if a user is in the rollout percentage using SHA-256 consistent hashing.
/// hash = SHA-256(utf8("{flagKey}:{userId}"))
/// value = BigEndian.Uint32(hash[0:4]) % 100
/// result = value < percentage
bool isInRollout(String flagKey, String userId, int percentage) {
  final input = utf8.encode('$flagKey:$userId');
  final hash = sha256.convert(input);
  final bytes = hash.bytes;
  final value = (bytes[0] << 24) | (bytes[1] << 16) | (bytes[2] << 8) | bytes[3];
  // Ensure unsigned comparison
  final unsigned = value & 0xFFFFFFFF;
  return (unsigned % 100) < percentage;
}

Object? getAttributeValue(String attribute, UserContext? user) {
  if (user == null) return null;
  switch (attribute) {
    case 'id':
      return user.id;
    case 'email':
      return user.email;
    default:
      return user.attributes[attribute];
  }
}

bool matchesCondition(String op, Object? attrValue, String condValue, UserContext? user) {
  final exists = attrValue != null && _toString(attrValue).isNotEmpty;

  switch (op) {
    case 'is_set':
      return exists;
    case 'is_not_set':
      return !exists;
  }

  if (!exists) return false;

  final value = _toString(attrValue).toLowerCase();
  final cv = condValue.toLowerCase();

  switch (op) {
    case 'equals':
    case 'eq':
      return value == cv;
    case 'not_equals':
    case 'neq':
      return value != cv;
    case 'contains':
      return value.contains(cv);
    case 'not_contains':
      return !value.contains(cv);
    case 'starts_with':
      return value.startsWith(cv);
    case 'ends_with':
      return value.endsWith(cv);
    case 'in':
      return _splitAndTrim(condValue).any((v) => v.toLowerCase() == value);
    case 'not_in':
      return !_splitAndTrim(condValue).any((v) => v.toLowerCase() == value);
    case 'greater_than':
    case 'gt':
      return _compareNumeric(attrValue, condValue, '>');
    case 'greater_equal':
    case 'gte':
      return _compareNumeric(attrValue, condValue, '>=');
    case 'less_than':
    case 'lt':
      return _compareNumeric(attrValue, condValue, '<');
    case 'less_equal':
    case 'lte':
      return _compareNumeric(attrValue, condValue, '<=');
    case 'regex':
      return _tryRegex(_toString(attrValue), condValue);
    case 'semver_gt':
      return _compareSemver(_toString(attrValue), condValue, '>');
    case 'semver_lt':
      return _compareSemver(_toString(attrValue), condValue, '<');
    case 'semver_eq':
      return _compareSemver(_toString(attrValue), condValue, '=');
    default:
      return false;
  }
}

String _toString(Object? v) {
  if (v == null) return '';
  if (v is String) return v;
  if (v is bool) return v ? 'true' : 'false';
  if (v is num) {
    // Format without trailing .0 for integers
    if (v == v.toInt()) return v.toInt().toString();
    return v.toString();
  }
  return v.toString();
}

List<String> _splitAndTrim(String s) =>
    s.split(',').map((p) => p.trim()).toList();

bool _compareNumeric(Object? attrVal, String condVal, String op) {
  final a = _toDouble(attrVal);
  if (a == null) return false;
  final b = double.tryParse(condVal);
  if (b == null) return false;

  switch (op) {
    case '>':
      return a > b;
    case '>=':
      return a >= b;
    case '<':
      return a < b;
    case '<=':
      return a <= b;
    default:
      return false;
  }
}

double? _toDouble(Object? v) {
  if (v == null) return null;
  if (v is num) return v.toDouble();
  if (v is String) return double.tryParse(v);
  return null;
}

bool _tryRegex(String value, String pattern) {
  try {
    return RegExp(pattern).hasMatch(value);
  } catch (_) {
    return false;
  }
}

bool _compareSemver(String attrVal, String condVal, String op) {
  final a = _parseVersion(attrVal);
  final b = _parseVersion(condVal);
  if (a == null || b == null) return false;

  while (a.length < b.length) a.add(0);
  while (b.length < a.length) b.add(0);

  for (int i = 0; i < a.length; i++) {
    if (a[i] > b[i]) return op == '>' || op == '>=';
    if (a[i] < b[i]) return op == '<' || op == '<=';
  }
  return op == '=' || op == '>=' || op == '<=';
}

List<int>? _parseVersion(String v) {
  final clean = v.startsWith('v') ? v.substring(1) : v;
  final parts = clean.split('.');
  final result = <int>[];
  for (final p in parts) {
    final n = int.tryParse(p);
    if (n == null) return null;
    result.add(n);
  }
  return result;
}
